src/libraries/Common/src/System/Net/Http/Http3/QPack/QPackEncoder.cs

```diff
@@ -0,0 +1,424 @@
+        {
+            Span<byte> buffer = stackalloc byte[IntegerEncoder.MaxInt32EncodedLength];
+
+            bool res = EncodeStaticIndexedHeaderField(index, buffer, out int bytesWritten);
+            Debug.Assert(res == true);
+
+            return buffer.Slice(0, bytesWritten).ToArray();
+        }
+
+        // https://tools.ietf.org/html/draft-ietf-quic-qpack-11#section-4.5.4
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | 0 | 1 | N | T |Name Index (4+)|
+        // +---+---+---+---+---------------+
+        // | H |     Value Length (7+)     |
+        // +---+---------------------------+
+        // |  Value String (Length bytes)  |
+        // +-------------------------------+
+        //
+        // Note for this method's implementation of above:
+        // - N is constant 0 here, indicating intermediates (proxies) can compress the header when fordwarding.
+        // - T is constant 1 here, indicating a static table reference.
+        // - H is constant 0 here, as we do not yet perform Huffman coding.
+        public static bool EncodeLiteralHeaderFieldWithStaticNameReference(int index, string value, Span<byte> destination, out int bytesWritten)
+        {
+            // Requires at least two bytes (one for name reference header, one for value length)
+            if (destination.Length >= 2)
+            {
+                destination[0] = 0b01010000;
+                if (IntegerEncoder.Encode(index, 4, destination, out int headerBytesWritten))
+                {
+                    destination = destination.Slice(headerBytesWritten);
+
+                    if (!EncodeValueString(value, destination, out int valueBytesWritten))
+                    {
+                        bytesWritten = headerBytesWritten + valueBytesWritten;
+                        return true;
+                    }
+                }
+            }
+
+            bytesWritten = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Encodes just the name part of a Literal Header Field With Static Name Reference. Must call <see cref="EncodeValueString(string, Span{byte}, out int)"/> after to encode the header's value.
+        /// </summary>
+        public static byte[] EncodeLiteralHeaderFieldWithStaticNameReferenceToArray(int index)
+        {
+            Span<byte> temp = stackalloc byte[IntegerEncoder.MaxInt32EncodedLength];
+
+            temp[0] = 0b01110000;
+            bool res = IntegerEncoder.Encode(index, 4, temp, out int headerBytesWritten);
+            Debug.Assert(res == true);
+
+            return temp.Slice(0, headerBytesWritten).ToArray();
+        }
+
+        public static byte[] EncodeLiteralHeaderFieldWithStaticNameReferenceToArray(int index, string value)
+        {
+            Span<byte> temp = value.Length < 256 ? stackalloc byte[256 + IntegerEncoder.MaxInt32EncodedLength * 2] : new byte[value.Length + IntegerEncoder.MaxInt32EncodedLength * 2];
```

Comment:
We don't want to use the array pool here for values >= 256?  Is it just not common enough to be worth it?

Reply:
These `ToArray` are called to create statics or per-pool, not per-request.